Documentation/design-docs/raw-eventlistener.md

```diff
@@ -0,0 +1,81 @@
+# Raw EventListener API #
+
+The goal of this design document is to describe what a Raw EventListener API will look like and how it will work.
+
+A raw API is one that dispatches events, but does not decode the data payload.  It simply provides a raw blob and a metadata blob that can be used to interpret the data when desired.
+
+## API Surface ##
+
+The existing EventListener API usage pattern is to:
+
+1. Create a new class that derives from EventListener.
+2. Implement the OnEventSourceCreated method to get callbacks when new EventSource objects are created (this allows EventListeners to subscribe to events).
+3. Implement the OnEventWritten method to get callbacks whenever a subscribed event is dispatched.
+
+The proposed extension of this pattern is the following:
+
+1. Add a new EventListener constructor that takes an EventListenerSettings enum parameter:
```

Comment:
Note that for over a decade we have wanted to fix a problem where we have an ugly issue that @anurse  pointed out (that you get callbacks for existing eventSources BEFORE your subclass of EventListener is fully initialized).   It is quite a pain.    Basicaly we need to start EventListener with event callbacks turned OFF, and then have a explict operation that turns them on.    

We can do this by having a new 'Start' method that takes your flags (I have added one)
* EventSourceCreated
* EventWritten
* EventWrittenSerialized (I have renamed it from 'raw')

THen have a new constructor
```
EventListener(bool startExplictly)
```
which turns off all events by default at start.     This solves the ugly 'not completely initialized' problem.   

P.S: in the past we also toyed with the idea of having events (that you subscribe to) instead of virtuals that you override.  This makes it is a bit easier to use (you don't need to make subclasses) and also solves the 'callback when not completely intialized problem.   If there is a desire to take that direction, I am open to that as well.  


We would alos need

Reply:
The way I have traditionally solved these design problems in other systems is by creating a TransactionManager interface that implements IDisposable and lets you arbitrarily nest transactions:

```
using (TransactionManager.OpenScope(_ => onEventSourceCreated, __ => onEventWritten, ___ => onEventWrittenSerialized )) // Resource acquisition is initialization (RAII)
{

}
```